CaT/src/TableRelations/Graphs/Graph.php

```diff
@@ -0,0 +1,252 @@
+		$from_id = $edge->fromId();
+		$to_id = $edge->toId();
+		return $this->checkNoIncommingOnlyConnections($from_id) && $this->checkNoIncommingOnlyConnections($to_id);
+	}
+
+
+	protected function checkNoCurrentSymmetricConnections($node_id) {
+		foreach ($this->neighbours($node_id) as $neighbour_id) {
+			if($this->edgeBetween($neighbour_id, $node_id)) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	protected function checkNoIncommingOnlyConnections($node_id) {
+		$neighbours_from = $this->neighbours($node_id);
+		$neighbours_to = array();
+		foreach ($this->connections as $a_node_id => $s_g_connections) {
+			foreach ($s_g_connections as $s_g => $to_ids) {
+				if(isset($to_ids[$node_id])) {
+					$neighbours_to[] = $a_node_id;
+				}
+			}
+		}
+		if(count(array_diff($neighbours_to,$neighbours_from))>0) {
+			return false;
+		}
+		return true;
+	}
+
+
+	protected function connectNodeIdsDirected($from_id, $to_id, AbstractEdge $edge) {
+		if($from_id === $to_id) {
+			throw new GraphException("cant't connect $from_id to itself");
+		}
+		if(!isset($this->nodes[$from_id]) || !isset($this->nodes[$to_id])) {
+			throw new GraphException("$from_id or $to_id does not exist in graph");
+		}
+
+		$to_subgraph = $this->getSubgraphOfNodeId($to_id);
+		if(isset($this->connections[$from_id][$to_subgraph][$to_id])) {
+			throw new GraphException("$from_id and $to_id allready connected");
+		}
+		if(isset($this->connections[$from_id])) {
+			if(isset($this->connections[$from_id][$to_subgraph])) {
+				$this->connections[$from_id][$to_subgraph][$to_id] = $edge;
+			} else {
+				$this->connections[$from_id][$to_subgraph] = array($to_id => $edge);
+			}
+		}
+		else {
+			$this->connections[$from_id] = array($to_subgraph => array($to_id => $edge));
+		}
+	}
+
+	public function nodeSubgraphId(AbstractNode $node) {
+		return $this->getSubgraphOfNodeId($node->id());
+	}
+
+	public function getSubgraphOfNodeId($node_id) {
+		return $this->nodes[$node_id]->subgraph();
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function getNodesBetween($from_id, $to_id, $subgraph = null) {
+		$return = array();
+		foreach ($this->DFS($from_id, $to_id, $subgraph) as $path) {
+			foreach ($path->sequence() as $node) {
+				$return[$node->id()] = $node;
+			}
+		}
+		return array_values($return);
+	}
+
+	public function getPathsBetween($from_id, $to_id, $sg = null) {
+		return $this->DFS($from_id, $to_id, $sg);
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function getNodesWithinSubgraphBetween($from_id, $to_id, $subgraph) {
+		return $this->getNodesBetween($from_id, $to_id, $subgraph);
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function connected($from_id, $to_id) {
+		return count($this->DFS($from_id, $to_id)) > 0;
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function nodes() {
+		return $this->nodes;
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function edges() {
+		return $this->edges;
+	}
+
+	public function edgeBetween($from_id, $to_id) {
```

Comment:
* Doku bitte.
* Muss das wirklich public sein?
* Fehlt das vielleicht in irgendeinem Interface?